lib/balance-tx/test/spec/Internal/Cardano/Write/Tx/BalanceSpec.hs

```diff
@@ -2203,25 +2198,36 @@ instance Arbitrary (MixedSign Value) where
         genPositive = arbitrary
     shrink (MixedSign v) = MixedSign <$> shrink v
 
-instance forall era. IsRecentEra era => Arbitrary (PartialTx era) where
-    arbitrary = do
-        tx <- genTxForBalancing
-        extraUTxO <- genExtraUTxO (txInputs tx)
-        let redeemers = []
-        let timelockKeyWitnessCounts = mempty
-        pure PartialTx {tx, extraUTxO, redeemers, timelockKeyWitnessCounts}
-      where
-        genExtraUTxO :: Set TxIn -> Gen (UTxO era)
-        genExtraUTxO = fmap UTxO . genMapFromKeysWith genTxOut
-        txInputs :: Tx era -> Set TxIn
-        txInputs tx = tx ^. bodyTxL . inputsTxBodyL
-    shrink partialTx@PartialTx {tx, extraUTxO} =
-        [ partialTx {extraUTxO = extraUTxO'}
-        | extraUTxO' <- shrinkInputResolution extraUTxO
-        ] <>
-        [ restrictResolution (partialTx {tx = tx'})
-        | tx' <- shrinkTx tx
-        ]
+instance IsRecentEra era => Arbitrary (PartialTx era) where
+    arbitrary = partialTxFromTxWithUTxO <$> genTxWithUTxO
+    shrink = shrinkMapBy
+        partialTxFromTxWithUTxO
+        txWithUTxOFromPartialTx
+        shrinkTxWithUTxO
+
+partialTxFromTxWithUTxO :: IsRecentEra era => TxWithUTxO era -> PartialTx era
+partialTxFromTxWithUTxO (TxWithUTxO tx extraUTxO) =
+    PartialTx {tx, extraUTxO, redeemers, timelockKeyWitnessCounts}
+  where
+    -- This embedding uses the following constants:
+    redeemers = []
+    timelockKeyWitnessCounts = mempty
+
+txWithUTxOFromPartialTx :: IsRecentEra era => PartialTx era -> TxWithUTxO era
+txWithUTxOFromPartialTx PartialTx {tx, extraUTxO} =
```

Comment:
This could also be done while retaining the use of named field puns. For example:

```hs
  txWithUTxOFromPartialTx
      PartialTx
          { tx
          , extraUTxO
          , redeemers                = _forgetThisField0
          , timelockKeyWitnessCounts = _forgetThisField1
          } =
          TxWithUTxO.constructFiltered
              tx
              extraUTxO
```

Although I'm not sure if this is really worth it, I do see your point.

In any case, I suspect we'll revisit this function definition very soon, so we can revise if necessary.

Reply:
Ah, but also: with "positional patterns" we guarantee that all fields have to be listed. I.e. we seeing `(PartialTx tx extraUTxO _redeemers _wits)` we can trust that it's only ignoring `_redeemers` and `_wits`, and that no one has added a new field `foo` to `PartialTx` without concidering whether it should be ignored or not in this function.

I think this would be my one concern in general with named field puns.